### ISSUE TITLE
no-useless-escapes has false positives for periods and meta sequences.

### ISSUE THREAD

**Issue by LJNeon** [eslint/eslint#9564]
**Tell us about your environment**

* **ESLint Version:** 4.10.0
* **Node Version:** 9.0.0
* **npm Version:** 5.5.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default
**Please show your full configuration:**
<details>
<summary>Configuration</summary>

```js
{
	"env": {
		"es6": true,
		"node": true
	},
	"extends": "eslint:recommended",
	"rules": {
		"indent": [2, "tab"],
		"linebreak-style": [2, "unix"],
		"quotes": [2, "double"],
		"semi": [2, "always"]
	},
	"parserOptions": {
		"sourceType": "module",
		"ecmaVersion": 2017
	}
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**
```js
this.regexes = {
	suffix: new RegExp("^(\d+(\.\d*)?|\.\d+)(e[+-]?\d+)?(" + this.suffixes.join("|") + ")?$", "i"),
	suffix_strict: new RegExp("^(\d+(\.\d*)?|\.\d+)(e[+-]?\d+)?(" + this.suffixes.join("|") + ")$", "i")
};
```
```bash
eslint src/**
```

**What did you expect to happen?**
It wouldn't complain about correct regex.
**What actually happened? Please include the actual, raw output from ESLint.**
It claimed the escapes were useless when they were needed.

**Comment by LJNeon** [comment]
```
   3:1  error  Expected indentation of 2 tabs but found 1  indent
   4:1  error  Expected indentation of 2 tabs but found 1  indent
   5:1  error  Expected indentation of 2 tabs but found 1  indent
   6:1  error  Expected indentation of 2 tabs but found 1  indent
   7:1  error  Expected indentation of 2 tabs but found 1  indent
   8:1  error  Expected indentation of 2 tabs but found 1  indent
   9:1  error  Expected indentation of 2 tabs but found 1  indent
  10:1  error  Expected indentation of 2 tabs but found 1  indent
  11:1  error  Expected indentation of 2 tabs but found 1  indent
  12:1  error  Expected indentation of 2 tabs but found 1  indent
  13:1  error  Expected indentation of 2 tabs but found 1  indent
  14:1  error  Expected indentation of 2 tabs but found 1  indent
  15:1  error  Expected indentation of 2 tabs but found 1  indent
  16:1  error  Expected indentation of 2 tabs but found 1  indent
  17:1  error  Expected indentation of 2 tabs but found 1  indent
  18:1  error  Expected indentation of 2 tabs but found 1  indent
  19:1  error  Expected indentation of 2 tabs but found 1  indent
  20:1  error  Expected indentation of 1 tab but found 0   indent
  21:1  error  Expected indentation of 2 tabs but found 1  indent
  22:1  error  Expected indentation of 2 tabs but found 1  indent
  23:1  error  Expected indentation of 2 tabs but found 1  indent
  24:1  error  Expected indentation of 1 tab but found 0   indent
```
here's the eslint output

**Comment by platinumazure** [comment]
I asked for more info to be absolutely sure, but I can give you one pointer right away.

Using a regexp literal vs invoking the `RegExp` constructor requires different "regular expression" syntax. The key is understanding that you're passing a string to the `RegExp` constructor, so the string escaping rules apply here.

Thus for this regex: `new RegExp("^(\d+(\.\d*)?|\.\d+)(e[+-]?\d+)?(" + this.suffixes.join("|") + ")$", "i")`

The backslash in all of the `\d` are being treated as string escapes, and `\d` (for example) is not a valid string escape, like `\n` would be. Your goal, presumably, is to match on digits in those areas; however, in order for that to happen, the backslash must be interpreted literally in the string. Thus you have to double up the backslash.

The rule is calling out that these escapes don't do what you think they are doing. In a string, `\d` is interpreted as `d`, which is not what you want.

Correct syntax: `new RegExp("^(\\d+(\\.\\d*)?|\\.\\d+)(e[+-]?\\d+)?(" + this.suffixes.join("|") + ")$", "i")`

**Comment by platinumazure** [comment]
@LJNeon Thanks, but I think you posted the output from your indent issue. This issue is about useless escapes. I did my best to guess at the answer to your question in my answer above-- if you don't feel the answer is complete, please include your no-useless-escape output. :smile:

**Comment by LJNeon** [comment]
```
   9:26  error  Unnecessary escape character: \d  no-useless-escape
   9:30  error  Unnecessary escape character: \.  no-useless-escape
   9:32  error  Unnecessary escape character: \d  no-useless-escape
   9:38  error  Unnecessary escape character: \.  no-useless-escape
   9:40  error  Unnecessary escape character: \d  no-useless-escape
   9:51  error  Unnecessary escape character: \d  no-useless-escape
  10:33  error  Unnecessary escape character: \d  no-useless-escape
  10:37  error  Unnecessary escape character: \.  no-useless-escape
  10:39  error  Unnecessary escape character: \d  no-useless-escape
  10:45  error  Unnecessary escape character: \.  no-useless-escape
  10:47  error  Unnecessary escape character: \d  no-useless-escape
  10:58  error  Unnecessary escape character: \d  no-useless-escape
```
here's the actual eslint output lmao, mixed up the two issues I opened.

**Comment by LJNeon** [comment]
and yeah that was the problem, thanks for clarifying.
